### ISSUE TITLE
CiliumEndpoints missing for Pods

### ISSUE THREAD

**Issue by dctrwatson** [cilium/cilium#27123]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Operations:
- pod is deleted
- pod is recreated (with same name)
- cep for pod is deleted

Pod no longer has an identity so its traffic is considered `world` and thus has no labels.

This happens randomly when there are dozens of recreations per second in the local cluster (hundreds in the clustermesh)

### Cilium Version

Client: 1.13.4 4061cdfc 2023-06-14T03:37:39+00:00 go version go1.19.10 linux/amd64
Daemon: 1.13.4 4061cdfc 2023-06-14T03:37:39+00:00 go version go1.19.10 linux/amd64

### Kernel Version

Linux ip-10-194-100-11 5.15.0-1039-aws #44~20.04.1-Ubuntu SMP Thu Jun 22 12:21:12 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.24.15-eks-a5565ad

### Anything else?

Seems awfully similar to https://github.com/cilium/cilium/issues/17047

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Thanks for this report @dctrwatson, that does seem not great, would you be able to grab a sysdump file and upload it please? (Using https://docs.cilium.io/en/stable/operations/troubleshooting/#automatic-log-state-collection )?

**Comment by dctrwatson** [comment]
> Thanks for this report @dctrwatson, that does seem not great, would you be able to grab a sysdump file and upload it please?

Unfortunately, I cannot without sanitizing certain outputs from it. If you know what parts of the sysdump that would be helpful for this issue, it would make it easier for me to share.

**Comment by dylandreimerink** [comment]
> Unfortunately, I cannot without sanitizing certain outputs from it

You can also send the sysdump via a non-public channel if that helps. Though I understand if you can't due to security concerns. 

> If you know what parts of the sysdump that would be helpful for this issue, it would make it easier for me to share.

Lets start with the following files:
* logs-cilium-*-cilium-agent.log
* k8s-pods-*.yaml
* ciliumendpoints-*.yaml
* k8s-events-*.yaml
* cilium-bugtool-cilium-*/cmd/cilium-endpoint-list.md
* cilium-bugtool-cilium-*/cmd/cilium-identity-list.md
* cilium-bugtool-cilium-*/cmd/cilium-map-events-cilium_ipcache--o-json.md

It would also be helpful if you could mention one or more of the pods that are effected so we can narrow down the search in the sysdump files.

**Comment by Sinjo** [comment]
Hey @dylandreimerink! I'm one of @dctrwatson's coworkers and I spent some time this week diving deeper into this.

I don't have a sysdump for you, but I've been digging through the Kubernetes apiserver audit logs and I do think I have the series of events that cause this issue for us.

# Our working theory

The problem seems to be a race condition that involves three controllers: Cilium agent, the Kubernetes core garbage collector, and our own custom operator that we use to schedule database pods.

It turns out that Cilium does correctly create the `CiliumEndpoint` (I checked that it was Cilium's service account doing this), but then the Kubernetes garbage collector ends up deleting it (I also checked the service account on this one and it was `system:controller:generic-garbage-collector`).

The series of events in the apiserver audit logs looks like this:

- Our operator deletes a database pod with background propagation of the deletion (`DeletePropagationBackground`) to child resources (which includes the `CiliumEndpoint`)
    - From reading the docs, background deletion is also the default for `kubectl`, so we're not doing anything particularly out of the ordinary here
    - This means that deletion of the child `CiliumEndpoint` will happen asynchronously
    - It's important to note that Cilium endpoints always have the same name as their corresponding pod, and we recreate database pods with consistent names when they're being restarted.  This means that a delayed delete can end up deleting the wrong Cilium endpoint (i.e. one that’s just been recreated by Cilium agent for the new pod that our operator created)
- Cilium agent deletes the corresponding endpoint for the old pod
- We create the new database pod
- Cilium agent tries to get the endpoint and finds nothing (I think this is part of its process of deciding whether to create or update an endpoint)
- Cilium agent creates the endpoint for the new pod
- The Kubernetes garbage collector runs, and starts by getting the endpoint by name, which succeeds (because it finds the new one Cilium just created)
- The Kubernetes garbage collector deletes the endpoint, thinking it was cleaning up the old one when it’s really deleting the new one that Cilium agent just created!
- We have a pod without a `CiliumEndpoint`, and we have to intervene manually by restarting it
    - In theory it could go wrong again when restarting it, but the window for this race condition seems fairly tight, so it doesn’t happen most of the time

I have copies of the specific logs from the apiserver, and I can get any extra info you'd like from them.

# Our current proposed workaround

We've been talking about what we can do on our end and we think the right answer for now is to have our controller delete the `CiliumEndpoint` before it deletes the pod. That would avoid the background-cascaded delete from ever being created and should prevent the race condition from happening.

# Longer term fix

If possible, we'd love to see Cilium handle this situation. It seems likely that this has happened elsewhere in the wild, but given that we only see it happen to a handful of pods when we roll the fleet, most people probably don't have enough workloads running to consistently trigger it. The race condition seems to happen within a window of 10-20ms, so I can see why it goes right most of the time.

I'm not familiar enough with the Cilium codebase to suggest how this could be fixed, but I'm hoping there's something reasonably easy you can do on your side!

Looking forward to hearing what you think.

**Comment by Sinjo** [comment]
Hey, I wanted to give you an update on this as we've been running that workaround for about a week now. Tl;dr: sadly it doesn't work, but we've got a new line of thinking that we're exploring because of what we've seen with it in place.

With the workaround in place, the series of events we see at the Kubernetes apiserver is:

- Our operator deletes the `CiliumEndpoint` for the pod we're about to delete (successful)
- Cilium tries to delete the endpoint in response to us deleting the pod (fails 404 because we already deleted it)
- Cilium does a get on the same name (also 404, so at this point the series of events is the same as before we introduced the workaround)
- Cilium creates a new endpoint with the same name, for the pod we just recreated (successful)
- Kube system garbage collector gets the endpoint (successful as cilium has recreated it)
- Kube system garbage collector deletes the new endpoint

This rules out my theory about it being a race between Cilium and the Kubernetes system garbage collector. With that out of the way, I spent some time looking at the `kube-controller-manager` logs and found something interesting (logs have been lightly redacted):

```
I0926 15:07:15.515291      11 garbagecollector.go:496] "Processing object" object="some-namespace/database-pod-123" objectUID=some-uid-123 kind="Pod" virtual=true
I0926 15:07:15.519930      11 garbagecollector.go:496] "Processing object" object="some-namespace/database-pod-123" objectUID=some-uid-456 kind="CiliumEndpoint" virtual=false
I0926 15:07:15.522892      11 garbagecollector.go:608] "Deleting object" object="some-namespace/database-pod-123" objectUID=some-uid-456 kind="CiliumEndpoint" propagationPolicy=Background
```

What stood out to me is that the garbage collector is processing the pod as a virtual node. If I understand right, this could imply that the Cilium agent is creating the new endpoint with an incorrect `ownerRef`. I'm very far from certain, but one thing I'm currently questioning is whether [the call to `GetPod` here](https://github.com/cilium/cilium/blob/e1eb268701d71ada610e8b14181714b757c2c64a/pkg/k8s/watchers/endpointsynchronizer.go#L155-L180) could be returning stale cached info, resulting in Kubernetes immediately deleting the endpoint because it references the old (now deleted) pod as its owner.

If I'm right in this line of thinking, it should only be possible for this to affect StatefulSets or anything that behaves similarly to them, where pod names remain consistent during rolling restarts (as is true in our operator).

We're very actively working on figuring this out, so I'll keep you updated with anything more that we learn.

**Comment by Sinjo** [comment]
Hello again 👋🏻

I've got a small update courtesy of something @dctrwatson noticed while debugging towards the end of last week.

It looks like whenever we see pods/endpoints reach this state, the agent on the corresponding node has internally deadlocked in some way. We think that might be the reason that the endpoint is created with stale data and why [the 10s reconcile loop](https://github.com/cilium/cilium/blob/207eb7a466244cbc06a50f4b7ecffc2fda08c3da/pkg/k8s/watchers/endpointsynchronizer.go#L99) never runs to recreate the endpoint after the Kubernetes GC deletes it.

While we couldn't find any reference to a fix for this specific issue in the release notes, we're aware that a decent amount of the relevant parts of the agent have been reworked between 1.13 and 1.14, so we're in the process of making that upgrade. We'll check back in once it's rolled out and we've had enough time to either run into this issue or for 1.14 to have solved it for us.

Thanks again for checking in on Slack. Hopefully this ends up being a non-issue after the upgrade 🤞🏻

**Comment by christarazi** [comment]
This sounds very similar to 

* https://github.com/cilium/cilium/issues/19931
* https://github.com/cilium/cilium/issues/17047

which should've been fixed. If it's true that the `ownerReference` that's set by K8s is somehow incorrect, then I can see how that would cause problems again.

**Comment by kahirokunn** [comment]
I have a same issue.

**Comment by kahirokunn** [comment]
Perhaps this is the same problem
https://github.com/cilium/cilium/issues/28303

**Comment by kahirokunn** [comment]
In my case, there are traces of the cilium operator moving to another node.
I found that many cilium endpoints do not exist.

**Comment by atamgp** [comment]
I believe to see this too, more often in v1.14.x  then before.

After I delete a node, I see a lot of unmanaged pods, then node starts, cilium pods starts on that node and endpoitns are created.
So no more unmanaged, but after some seconds, It repeats...

used this script:
https://raw.githubusercontent.com/cilium/cilium/main/contrib/k8s/k8s-unmanaged.sh

I think this is the log, **Identity of endpoint changed** from **oldIdentity="no identity"** ?
What could be the issue here?

> cilium-6xbxb cilium-agent level=info msg="Identity of endpoint changed" containerID=735c1fc3ed datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=581 identity=19295 identityLabels="k8s:app.kubernetes.io/instance=capsule,k8s:app.kubernetes.io/managed-by=Helm,k8s:app.kubernetes.io/name=capsule,k8s:app.kubernetes.io/version=0.3.3,k8s:app=capsule,k8s:helm.sh/chart=capsule-0.4.5,k8s:io.cilium.k8s.namespace.labels.goldilocks.fairwinds.com/enabled=true,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=capsule-system,k8s:io.cilium.k8s.namespace.labels.pod-security.kubernetes.io/enforce=privileged,k8s:io.cilium.k8s.namespace.labels.pod-security.kubernetes.io/warn=baseline,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=capsule,k8s:io.kubernetes.pod.namespace=capsule-system" ipv4=100.64.187.194 ipv6="fe80::dcd0:caff:feeb:f2b1" k8sPodName=capsule-system/capsule-controller-manager-79d99b54d7-c859w oldIdentity="no identity" subsys=endpoint

**Comment by atamgp** [comment]
Some additional notes:

I tracked the cilium pod on the same node of an unmanaged pod
In the whole log of this cilium pod since start, there is no entry of that unmanaged pod's name..

The cilium pod has cilium-agent startedAt: "2023-11-28T11:14:46Z"
The unmanaged pod has state running startedAt: "2023-11-28T11:14:58Z"

So the unmanaged pod started 12 seconds after the agent on the node was ready, still neglected and no mention of.

Next, I delete the unmanaged pod and lucky, this was rescheduled on the same node, same cilium agent :)
Now a cep was created....

> ➜  deploy git:(cilium_bump) ✗ k -n kube-system logs cilium-6zf5h | grep 6qjjd                                                                                                                                                                                                                            <aws:branch-cluster>
> Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
> level=info msg="Create endpoint request" addressing="&{100.64.40.225   fe80::c47b:a3ff:feec:bb50  }" containerID=3d37d33057f6b41007819ecc305b6ef26bc038883fb9dec0ff102b0de82f991e datapathConfiguration="&{false true true true true 0xc0021c8be0}" interface=eni14e3e134742 k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd labels="[]" subsys=daemon sync-build=true
> level=info msg="New endpoint" containerID=3d37d33057 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2898 ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd subsys=endpoint
> level=info msg="Invalid state transition skipped" containerID=3d37d33057 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2898 endpointState.from=waiting-for-identity endpointState.to=waiting-to-regenerate file=/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd line=605 subsys=endpoint
> level=info msg="Waiting for endpoint to be generated" containerID=3d37d33057 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2898 ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd subsys=endpoint
> level=info msg="Resolving identity labels (blocking)" containerID=3d37d33057 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2898 identityLabels="k8s:app=gatemin-ext-auth-filter,k8s:io.cilium.k8s.namespace.labels.admission.gatekeeper.sh/ignore=true,k8s:io.cilium.k8s.namespace.labels.goldilocks.fairwinds.com/enabled=true,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=gatemin-system,k8s:io.cilium.k8s.namespace.labels.pod-security.kubernetes.io/enforce=privileged,k8s:io.cilium.k8s.namespace.labels.pod-security.kubernetes.io/warn=baseline,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=gatemin-ext-auth-filter,k8s:io.kubernetes.pod.namespace=gatemin-system" ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd subsys=endpoint
> level=info msg="Identity of endpoint changed" containerID=3d37d33057 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2898 identity=17297 identityLabels="k8s:app=gatemin-ext-auth-filter,k8s:io.cilium.k8s.namespace.labels.admission.gatekeeper.sh/ignore=true,k8s:io.cilium.k8s.namespace.labels.goldilocks.fairwinds.com/enabled=true,k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=gatemin-system,k8s:io.cilium.k8s.namespace.labels.pod-security.kubernetes.io/enforce=privileged,k8s:io.cilium.k8s.namespace.labels.pod-security.kubernetes.io/warn=baseline,k8s:io.cilium.k8s.policy.cluster=default,k8s:io.cilium.k8s.policy.serviceaccount=gatemin-ext-auth-filter,k8s:io.kubernetes.pod.namespace=gatemin-system" ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd oldIdentity="no identity" subsys=endpoint
> level=info msg="Rewrote endpoint BPF program" containerID=3d37d33057 datapathPolicyRevision=0 desiredPolicyRevision=20 endpointID=2898 identity=17297 ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd subsys=endpoint
> level=info msg="Successful endpoint creation" containerID=3d37d33057 datapathPolicyRevision=20 desiredPolicyRevision=20 endpointID=2898 identity=17297 ipv4=100.64.40.225 ipv6="fe80::c47b:a3ff:feec:bb50" k8sPodName=gatemin-system/gatemin-ext-auth-filter-65b6c6bb99-6qjjd subsys=daemon

**Comment by bumarcell** [comment]
@atamgp I believe `cilium endpoint regenerate <id>` would also do the trick (instead of deleting the pod)
What I noticed in my case (pod exists without cep) is that `cilium endpoint list` will show the endpoint in the list, but there's no cep object on k8s, so I took its ID and ran the command above, which I believe recreated the cep.

**Comment by kahirokunn** [comment]
Keep

**Comment by atamgp** [comment]
it seems to be fixed in 1.14.7?

**Comment by Sinjo** [comment]
I don't think we've seen it since we moved to 1.14. I've definitely not seen it in the last month or so.

I did take a look through our internal Slack for the last time this happened and found something I saw after my initial message above and forgot to share here. Whenever we saw the error, it was correlated with this in the logs:

```
level=warning msg="Status check error channel is full, dropping this error" error="quorum check failed 8578 times in a row: 95h50m38.701162167s since last heartbeat update has been received" subsys=kvstore
```

**Comment by atamgp** [comment]
aha thanks for sharing. 
I will move to latest 1.14 and see how that goes in prod.

**Comment by christarazi** [comment]
@dctrwatson Would you be able to try https://github.com/cilium/cilium/pull/31605 and see if this fixes your issue?

**Comment by kahirokunn** [comment]
Keep

**Comment by llhhbc** [comment]
> #31605

Hi, why not change ciliumendpoint's owner to ciliumnode, not to the pod. And I have found `doCiliumEndpointSyncGC` to gc ciliumEndpoint

**Comment by aanm** [comment]
> > [#31605](https://github.com/cilium/cilium/pull/31605)
> 
> [@Sinjo](https://github.com/Sinjo) [@christarazi](https://github.com/christarazi) Hi, why not change ciliumendpoint's owner to ciliumnode, not to the pod. And I have found `doCiliumEndpointSyncGC` to gc ciliumEndpoint

Because CiliumEndpoints need to be removed when a Pod is destroyed, otherwise there's a risk of overlapping endpoint IPs.

**Comment by aanm** [comment]
There has been a lot of fixes regarding CiliumEndpoints made into v1.15.5. Please upgrade to this version. If you still face any issues feel free to create a new GitHub issue. Thank you.
